dynamic_programming/hamiltons-cycle.cpp

```diff
@@ -0,0 +1,147 @@
+#include <iostream>
+#include <vector>
+
+typedef bool (*test_func_t)(void);
+
+using namespace std;
+
+
+/* function to takes boolean the matrix of
+    [i, j] is true if there is a road between from i to j */
+bool hamilton_cycle(bool **routes, int n){
+    int height = 1 << n, width = n;
+    bool dp[(1 << n)][n];
+    for (int i=0; i<height; i++){
+        for (int j=0; j<width; j++){
+            dp[i][j] = false;
+        }
+    }
+
+    dp[1][0] = true;
+    for (int i=1; i<height; i++){
+        vector <int> zeros, ones;
+        /* finding positions with 1s and 0s and separate them */
+        for (int pos=0; pos < width; ++pos){
+            if ((1 << pos) & i){
+                ones.push_back(pos);
+            } else {
+                zeros.push_back(pos);
+            }
+        }
+
+        for (int pos = 0; pos < ones.size(); ++pos){
+            int o = ones[pos];
+            if (!dp[i][o])
+                continue;
+
+            for (int inner_pos = 0; inner_pos < zeros.size(); inner_pos++){
+                int z = zeros[inner_pos];
+                if (!routes[o][z])
+                    continue;
```

Comment:
Make it look like C++ code.
```suggestion
                if (!routes[o][z]) {
                    continue;
                }
```